### PR TITLE
Remove unnecesary blocking operations

### PR DESCRIPTION
**Pull request by alcuadrado**
Third PR of the compilation times optimization effort.

This PR removes many blocking operations in favor of async ones. A common pattern here is changing an iteration for a `await Promise.all(arr.map(...))`.

I believe this PR is ready to review, as these are internal changes that probably don't require new tests.

Depends on https://github.com/NomicFoundation/hardhat/pull/3049

### CODE REVIEW

File: packages/hardhat-core/src/internal/artifacts.ts [hunk 1]
```diff
@@ -129,22 +129,24 @@ export class Artifacts implements IArtifacts {
...
-    });
-
-    if (pathToBuildInfo === undefined) {
-      return;
-    }
+    await Promise.all([
+      fsExtra.writeJSON(artifactPath, artifact, {
+        spaces: 2,
+      }),
+      async () => {
```
[the remark this answers is not in the document]
Comment (alcuadrado) [reply packages/hardhat-core/src/internal/artifacts.ts:136]: Great catch! thanks

File: packages/hardhat-core/src/internal/artifacts.ts [hunk 2]
```diff
@@ -180,12 +182,24 @@ export class Artifacts implements IArtifacts {
    * Remove all artifacts that don't correspond to the current solidity files
    */
   public async removeObsoleteArtifacts() {
-    const validArtifactsPaths = new Set<string>();
+    const validArtifactPaths = await Promise.all(
+      this._validArtifacts.map(({ sourceName, artifacts }) =>
```
[the remark this answers is not in the document]
Comment (alcuadrado) [reply packages/hardhat-core/src/internal/artifacts.ts:186]: I don't understand your comment tbh.

[the remark this answers is not in the document]
Comment (fvictorio) [reply packages/hardhat-core/src/internal/artifacts.ts:186]: This value is later used as this:

```js
    const validArtifactsPathsSet = new Set<string>(...validArtifactPaths);
```

And the type of `validArtifactPaths` is `string[][]`. This means that we are doing this:

```
new Set(validArtifactsPaths[0], validArtifactPaths[1], ...)
```

Where each one of those values is a `string[]`. But the `Set` constructor will only use the first one. I think this is a mistake.

**Comment by alcuadrado** [comment]
This PR should be ready

File: packages/hardhat-core/src/internal/artifacts.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (alcuadrado) [reply packages/hardhat-core/src/internal/artifacts.ts]: yep, I fixed it. thanks

**Review by fvictorio — APPROVED** [review]

**Pull request merged**